### PR TITLE
deb: test deb package by piuparts

### PR DESCRIPTION
**Pull request by kenhys**

### CODE REVIEW

**Comment by cosmo0920** [comment]
Due to this line:
https://github.com/fluent-plugins-nursery/td-agent-builder/blob/master/td-agent/apt/debian-buster/Dockerfile#L53
We need to execute `apt update` before installing package with apt.

**Comment by cosmo0920** [comment]
Oh, `piuparts`! I got it.

**Comment by kenhys** [comment]
```
Piuparts caught exception, exiting...
------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/sbin/piuparts", line 3260, in <module>
    main()
  File "/usr/sbin/piuparts", line 3252, in main
    process_packages(regular_packages_list)
  File "/usr/sbin/piuparts", line 3147, in process_packages
    chroot.create()
  File "/usr/sbin/piuparts", line 779, in create
    self.create_temp_dir()
  File "/usr/sbin/piuparts", line 770, in create_temp_dir
    self.name = tempfile.mkdtemp(dir=settings.tmpdir)
  File "/usr/lib/python2.7/tempfile.py", line 339, in mkdtemp
    _os.mkdir(file, 0700)
OSError: [Errno 30] Read-only file system: '/fluentd/td-agent/apt/repositories/debian/pool/buster/main/t/td-agent/tmpQSOYEP'
------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/sbin/piuparts", line 3274, in <module>
    panic(1)
  File "/usr/sbin/piuparts", line 490, in panic
```

fails.

**Comment by kenhys** [comment]
it may be succeed when not to use /fluentd/

**Comment by cosmo0920** [comment]
Yep. piuparts can be started with this commit:https://github.com/fluent-plugins-nursery/td-agent-builder/pull/164/commits/1fb99a5c69c9050882bf3aa0b2e368e11d6f504b

https://github.com/fluent-plugins-nursery/td-agent-builder/pull/164/checks?check_run_id=947401707#step:6:1687

**Comment by kenhys** [comment]
https://github.com/fluent-plugins-nursery/td-agent-builder/pull/164/checks?check_run_id=947531443

```
  E: Release signed by unknown key (key id 901F9177AB97ACBE)
     The specified keyring /usr/share/keyrings/debian-archive-keyring.gpg may be incorrect or out of date.
     You can find the latest Debian release key at https://ftp-master.debian.org/keys.html
```

**Comment by kenhys** [comment]
It may be caused by old keyring.
https://packages.ubuntu.com/search?suite=bionic&searchon=names&keywords=debian-archive-keyring bionic 2017.7ubuntu1: all 
https://packages.debian.org/search?suite=buster&searchon=names&keywords=debian-archive-keyring buster 2019.1: all

**Comment by kenhys** [comment]
> It may be caused by old keyring.
> https://packages.ubuntu.com/search?suite=bionic&searchon=names&keywords=debian-archive-keyring bionic 2017.7ubuntu1: all
> https://packages.debian.org/search?suite=buster&searchon=names&keywords=debian-archive-keyring buster 2019.1: all

1. fetch buster keyring package
2. --privileged mode?

**Comment by cosmo0920** [comment]
Really? `901F9177AB97ACBE` should not be included debian keyring.
It is Treasure Data key.

```
% gpg --search-keys 901F9177AB97ACBE                                                     
gpg: data source: http://keys.gnupg.net:11371
(1)	Treasure Data, Inc (Treasure Agent Official Signing key) <support@trea
	  4096 bit RSA key 901F9177AB97ACBE, 作成: 2016-12-27
Keys 1-1 of 1 for "901F9177AB97ACBE".  番号(s)、N)次、またはQ)中止を入力してください > q
gpg: error searching keyserver: 操作がキャンセルされました
```

```
% gpg --recv-key 901F9177AB97ACBE
gpg: 鍵901F9177AB97ACBE: 公開鍵"Treasure Data, Inc (Treasure Agent Official Signing key) <support@treasure-data.com>"をインポートしました
gpg: 処理数の合計: 1
gpg:               インポート: 1
```

**Comment by kenhys** [comment]
Oh, 

> Really? 901F9177AB97ACBE should not be included debian keyring.
> It is Treasure Data key.

I've overlooked it, thanks!

**Comment by kenhys** [comment]
So, I need to prepare Treasure Data key 
https://docs.fluentd.org/installation/install-by-deb

**Comment by kenhys** [comment]
curl https://packages.treasuredata.com/GPG-KEY-td-agent

**Comment by kenhys** [comment]
NOTE: apt-key will be removed in the future release 

```
apt (2.1.8) unstable; urgency=medium
    
     [ Julian Andres Klode ]
    * Fully deprecate apt-key, schedule removal for Q2/2022
```

buster is not affected, but bullseye (Debian 11), it will not available anymore.

**Comment by kenhys** [comment]
piuparts runnning with Treasure Data key, but it shoots eatmydata issue.

```
0m0.0s DEBUG: Starting command: ['debootstrap', '--variant=minbase', '--keyring=/usr/share/keyrings/td-agent-archive-keryring.gpg', '--include=eatmydata', '--no-merged-usr', '--components=buster,contrib', 'buster', '/tmp/tmp0W0uoB', 'http://packages.treasuredata.com/4/debian/buster/']
0m0.4s DUMP: 
  I: Target architecture can be executed
  I: Retrieving InRelease 
  I: Checking Release signature
  I: Valid Release signature (key id BEE682289B2217F45AF4CC3F901F9177AB97ACBE)
  I: Retrieving Packages 
  I: Validating Packages 
  I: Resolving dependencies of required packages...
  I: Resolving dependencies of base packages...
  I: Checking component contrib on http://packages.treasuredata.com/4/debian/buster...
  E: Couldn't find these debs: eatmydata apt
0m0.4s ERROR: Command failed (status=1): ['debootstrap', '--variant=minbase', '--keyring=/usr/share/keyrings/td-agent-archive-keryring.gpg', '--include=eatmydata', '--no-merged-usr', '--components=buster,contrib', 'buster', '/tmp/tmp0W0uoB', 'http://packages.treasuredata.com/4/debian/buster/']
  I: Target architecture can be executed
  I: Retrieving InRelease 
```

**Comment by kenhys** [comment]
It seems that official mirror is mising.

**Comment by kenhys** [comment]
```
Selecting previously unselected package libeatmydata1:amd64.
Preparing to unpack .../38-libeatmydata1_105-7_amd64.deb ...
Unpacking libeatmydata1:amd64 (105-7) ...
Selecting previously unselected package eatmydata.
Preparing to unpack .../39-eatmydata_105-7_all.deb ...
Unpacking eatmydata (105-7) ...
```

eatmydata is installed.

**Comment by kenhys** [comment]
piuparts(1) says: "Only the first mirror is used with debootstrap."

**Comment by kenhys** [comment]
```
0m41.3s DEBUG: Command ok: ['debootstrap', '--variant=minbase', '--keyring=/usr/share/keyrings/debian-archive-keyring.gpg', '--no-merged-usr', '--components=buster,main,main', 'buster', '/tmp/tmphFdPA0', 'http://deb.debian.org/debian']
0m41.3s DEBUG: Starting command: ['mount', '-t', 'proc', 'proc', '/tmp/tmphFdPA0/proc']
0m41.3s DUMP: 
  mount: /tmp/tmphFdPA0/proc: permission denied.
0m41.3s ERROR: Command failed (status=32): ['mount', '-t', 'proc', 'proc', '/tmp/tmphFdPA0/proc']
  mount: /tmp/tmphFdPA0/proc: permission denied.
```

found permission issue

**Comment by kenhys** [comment]
It requires --priviledged.

File: .github/workflows/apt.yml [hunk 1]
```diff
@@ -47,6 +47,7 @@ jobs:
       - name: Installation Test
         run: |
           docker run \
+          --priviledged \
```
Comment (cosmo0920) [inline .github/workflows/apt.yml:50]: Typo? 
`privileged`?

Comment (kenhys) [reply .github/workflows/apt.yml]: Yes.

**Comment by kenhys** [comment]
piuparts works (even though it fails)

```
0m43.4s DEBUG: Starting command: ['umount', '/tmp/tmpikq4F6/dev/shm']
0m43.4s DEBUG: Command ok: ['umount', '/tmp/tmpikq4F6/dev/shm']
0m43.4s DEBUG: Starting command: ['umount', '/tmp/tmpikq4F6/dev/console']
0m43.4s DEBUG: Command ok: ['umount', '/tmp/tmpikq4F6/dev/console']
0m43.4s DEBUG: Starting command: ['umount', '/tmp/tmpikq4F6/dev/ptmx']
0m43.4s DEBUG: Command ok: ['umount', '/tmp/tmpikq4F6/dev/ptmx']
0m43.4s DEBUG: Starting command: ['umount', '/tmp/tmpikq4F6/dev/pts']
0m43.4s DEBUG: Command ok: ['umount', '/tmp/tmpikq4F6/dev/pts']
0m43.4s DEBUG: Starting command: ['umount', '/tmp/tmpikq4F6/proc']
0m43.4s DEBUG: Command ok: ['umount', '/tmp/tmpikq4F6/proc']
0m43.4s DEBUG: Starting command: ['rm', '-rf', '--one-file-system', '/tmp/tmpikq4F6']
0m43.6s DEBUG: Command ok: ['rm', '-rf', '--one-file-system', '/tmp/tmpikq4F6']
0m43.6s DEBUG: Removed directory tree at /tmp/tmpikq4F6
0m43.6s ERROR: piuparts run ends.
##[error]Process completed with exit code 1.
```

**Comment by kenhys** [comment]
Use /etc/apt/trusted.gpg.d/.

**Comment by kenhys** [comment]
To enable additional keyring, use debootstrap in advance.

**Comment by kenhys** [comment]
```
   
  The following packages have unmet dependencies:
   td-agent : Depends: libyaml-0-2 but it is not installable
  E: Unable to correct problems, you have held broken packages.
0m5.2s ERROR: Command failed (status=100): ['chroot', '/tmp/tmpTYY4wQ', 'apt-get', '-y', '--allow-downgrades', 'install', './tmp/td-agent-dbgsym_4.0.0-1_amd64.deb', './tmp/td-agent_4.0.0-1_amd64.deb']
  Reading package lists...
  Building dependency tree...
  Some packages could not be installed. This may mean that you have
  requested an impossible situation or if you are using the unstable
  distribution that some required packages have not yet been created
  or been moved out of Incoming.
  The following information may help to resolve the situation:
  
  The following packages have unmet dependencies:
   td-agent : Depends: libyaml-0-2 but it is not installable
  E: Unable to correct problems, you have held broken packages.
```

It seems that keyring issue  was fixed.

**Comment by kenhys** [comment]
It will be fixed runtime dependency is installed in advance.

File: td-agent/apt/install-test.sh [hunk 2]
```diff
@@ -12,13 +12,34 @@ case ${code_name} in
...
 esac
 apt install -V -y \
   ${repositories_dir}/${distribution}/pool/${code_name}/${channel}/*/*/*_${architecture}.deb
 
 td-agent --version
+
+find ${repositories_dir}
+apt install -V -y piuparts mount gnupg1 curl eatmydata
+curl https://packages.treasuredata.com/GPG-KEY-td-agent > td-agent.gpg
+TD_AGENT_KEYRING=/usr/share/keyrings/td-agent-archive-keryring.gpg
```
Comment (cosmo0920) [inline td-agent/apt/install-test.sh:31]:
```diff
-keryring
+keyring
```
?

Comment (kenhys) [reply td-agent/apt/install-test.sh:31]: Ah, I'll fix it.

Comment (kenhys) [reply td-agent/apt/install-test.sh:31]: Thanks!

**Comment by kenhys** [comment]
It seems that `chroot /var/lib/chroot/buster-root/ apt install -y libyaml-0-2` will work.

**Comment by kenhys** [comment]
```
2020-08-07T02:12:16.8293296Z 0m14.8s DEBUG: Starting command: ['chroot', '/tmp/tmpxGttLP', '/usr/sbin/logrotate', '/etc/logrotate.d/td-agent']
2020-08-07T02:12:16.8311096Z 0m14.8s DUMP: 
2020-08-07T02:12:16.8311800Z   error: stat of /var/log/td-agent/td-agent.log failed: No such file or directory
```

lotating failure.

**Comment by cosmo0920** [comment]
piuparts still complains:

```
   The following packages have unmet dependencies:
   td-agent : Depends: libssl1.1 (>= 1.1.1) but 1.1.0g-2ubuntu4 is to be installed
  E: Unable to correct problems, you have held broken packages.
```

**Comment by cosmo0920** [comment]
And Ubuntu Xenial does not have gnupg1:
```
E: Unable to locate package gnupg1
##[error]Process completed with exit code 100.
```

File: td-agent/apt/install-test.sh [hunk 3]
```diff
@@ -26,7 +26,14 @@ apt install -V -y \
 td-agent --version
 
 find ${repositories_dir}
-apt install -V -y piuparts mount gnupg1 curl eatmydata
+case ${code_name} in
+    xenial)
+	apt install -V -y piuparts mount gnupg curl eatmydata
+	;;
+    *)
+	apt install -V -y piuparts mount gnupg curl eatmydata
```
Comment (cosmo0920) [inline td-agent/apt/install-test.sh:34]: `gnupg` means `gnupg1` in this line?

Comment (kenhys) [reply td-agent/apt/install-test.sh]: Oh, yes.

**Comment by cosmo0920** [comment]
Ubuntu Xenial does not provide gpg1. We can use gpg instead of gpg1:

ref: https://github.com/fluent-plugins-nursery/td-agent-builder/pull/164/checks?check_run_id=956865808#step:6:2517

```
 + gpg1 --no-default-keyring --keyring /usr/share/keyrings/td-agent-archive-keyring.gpg --import td-agent.gpg
/fluentd/td-agent/apt/install-test.sh: line 39: gpg1: command not found
```

**Comment by cosmo0920** [comment]
Ubuntu Focal jobs was stuck due to interactive selections:

```
debconf: falling back to frontend: Readline
Configuring tzdata
------------------

Please select the geographic area in which you live. Subsequent configuration
questions will narrow this down by presenting a list of cities, representing
the time zones in which they are located.

  1. Africa      4. Australia  7. Atlantic  10. Pacific  13. Etc
  2. America     5. Arctic     8. Europe    11. SystemV
  3. Antarctica  6. Asia       9. Indian    12. US
##[error]The operation was canceled.
```

ref: https://github.com/fluent-plugins-nursery/td-agent-builder/runs/956381796?check_suite_focus=true#step:6:1953

How can we solve this?

**Comment by kenhys** [comment]
```
0m15.2s ERROR: FAIL: Package purging left files on system:
  /etc/td-agent/	 owned by: td-agent
  /etc/td-agent/plugin/	 not owned
  /var/log/td-agent/	 not owned
```
above issue was fixed.

**Comment by kenhys** [comment]
```
0m17.7s ERROR: FAIL: After purging files have disappeared:
  /opt/	 owned by: td-agent
```

above error still happens. It is strage.

**Comment by kenhys** [comment]
`mkdir -p /opt/ in postrm` pass piuparts test , but I'm no willing to do it.

**Comment by kenhys** [comment]
```
2020-08-11T02:44:28.1562383Z + ls /var/lib/chroot/buster-root/
2020-08-11T02:44:28.1573877Z bin   dev  home  lib32	libx32	mnt  proc  run	 srv  tmp  var
2020-08-11T02:44:28.1577642Z boot  etc  lib	 lib64	media	opt  root  sbin  sys  usr
2020-08-11T02:44:28.1578501Z + piuparts --distribution=buster --existing-chroot=/var/lib/chroot/buster-root --keyring=/usr/share/keyrings/td-agent-archive-keyring.gpg '--mirror=http://packages.treasuredata.com/4/debian/buster/ buster contrib' --skip-logrotatefiles-test --warn-on-leftovers-after-purge /tmp/td-agent-dbgsym_4.0.0-1_amd64.deb /tmp/td-agent_4.0.0-1_amd64.deb
```

before runnning piuparts, opt is already created.

**Comment by kenhys** [comment]
* bionic debootstrap: security update is not applied?

**Comment by kenhys** [comment]
On bionic chroot, there is no security updates:

```
  # cat chroot/bionic-root/etc/apt/sources.list 
  deb http://archive.ubuntu.com/ubuntu bionic main
```

**Comment by kenhys** [comment]
suppress manual configuration via noninteractive when tzdata is installed.

**Comment by kenhys** [comment]
@ashie could you review this PR, please?

File: td-agent/apt/install-test.sh [hunk 4]
```diff
@@ -31,6 +31,10 @@ case ${code_name} in
 	apt install -V -y piuparts mount gnupg curl eatmydata
 	gpg_command=gpg
 	;;
+    focal)
+	DEBIAN_FRONTEND=noninteractive apt install -V -y piuparts mount gnupg1 curl eatmydata
```
Comment (ashie) [inline td-agent/apt/install-test.sh:35]: I want to add `DEBIAN_FRONTEND=noninteractive` to all versions and merge it to `*)` block if it's harmless.

File: td-agent/templates/package-scripts/td-agent/deb/postrm [hunk 5]
```diff
@@ -6,16 +6,19 @@ set -e
 
 if [ "$1" = "purge" ]; then
         rm -f /etc/default/<%= project_name %>
-        rm -f /etc/<%= project_name %>/<%= project_name %>.conf
 	dpkg-statoverride --list /etc/<%= project_name %> > /dev/null && \
 		dpkg-statoverride --remove /etc/<%= project_name %>
-	rm -f /var/run/<%= project_name %>/*
+        rm -f /etc/<%= project_name %>/<%= project_name %>.conf
```
[the remark this answers is not in the document]
Comment (kenhys) [reply td-agent/templates/package-scripts/td-agent/deb/postrm]: Thanks, I've aligned indent.

File: td-agent/apt/install-test.sh (hunk 4 above)
Comment (kenhys) [reply td-agent/apt/install-test.sh:35]: I'll check it again whether it is focal specific.

Comment (kenhys) [reply td-agent/apt/install-test.sh:35]: I've unified into `*)` block.
@ashie

**Pull request merged**